realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java

```diff
@@ -0,0 +1,297 @@
+                assertEquals(0, changes.getChanges().length);
+                looperThread.testComplete();
+            }
+        });
+        realm.beginTransaction();
+        realm.where(AllTypes.class).equalTo(AllTypes.FIELD_LONG, 0).findFirst().setColumnLong(10);
+        realm.where(AllTypes.class).equalTo(AllTypes.FIELD_LONG, 9).findFirst().setColumnLong(0);
+        realm.commitTransaction();
+    }
+
+    @Test
+    @RunTestInLooperThread
+    public void mixed_changes() {
+        Realm realm = looperThread.realm;
+        populateData(realm, 10);
+        RealmResults<AllTypes> results = realm.where(AllTypes.class).findAllSorted(AllTypes.FIELD_LONG);
+        results.addChangeListener(new OrderedRealmCollectionChangeListener<RealmResults<AllTypes>>() {
+            @Override
+            public void onChange(RealmResults<AllTypes> collection, OrderedCollectionChangeSet changes) {
+                checkRanges(changes.getDeletionRanges(),
+                        0, 2,
+                        5, 1);
+                assertArrayEquals(changes.getDeletions(), new long[]{0, 1, 5});
+
+                checkRanges(changes.getInsertionRanges(),
+                        0, 2,
+                        9, 2);
+                assertArrayEquals(changes.getInsertions(), new long[]{0, 1, 9, 10});
+
+                checkRanges(changes.getChangeRanges(),
+                        3, 2,
+                        8, 1);
+                assertArrayEquals(changes.getChanges(), new long[]{3, 4, 8});
+
+                looperThread.testComplete();
+            }
+        });
+
+        realm.beginTransaction();
+        createObjects(realm, 11, 12, -1, -2);
+        deleteObjects(realm, 0, 1, 5);
+        modifyObjects(realm, 12, 3, 4, 9);
+        realm.commitTransaction();
+        // After transaction, '*' means the object has been modified. 12 has been modified as well, but it is created
+        // and modified in the same transaction, should not be counted in the changes range.
+        // [-1, -2, 2, *3, *4, 6, 7, 8, *9, 11, 12]
+    }
+
+    // The change set should empty when the async query returns at the first time.
+    @Test
+    @RunTestInLooperThread
+    public void emptyChangeSet_findAllAsync(){
+        Realm realm = looperThread.realm;
+        populateData(realm, 10);
+        final RealmResults<AllTypes> results = realm.where(AllTypes.class).findAllSortedAsync(AllTypes.FIELD_LONG);
+        results.addChangeListener(new OrderedRealmCollectionChangeListener<RealmResults<AllTypes>>() {
+            @Override
+            public void onChange(RealmResults<AllTypes> collection, OrderedCollectionChangeSet changes) {
+                assertSame(collection, results);
+                assertEquals(9, collection.size());
+                assertNull(changes);
+                looperThread.testComplete();
+            }
+        });
+
+        final CountDownLatch bgDeletionLatch = new CountDownLatch(1);
+        new Thread(new Runnable() {
```

Comment:
// beginTransaction() will make the async query return immediately. So we have to create an object in another
        // thread. Also, the latch has to be counted down after transaction committed so the async query results can
        // contain the modification in the background transaction.

Added above comments in test.